lib/gcf_sms/gcf_sms.c

```diff
@@ -0,0 +1,479 @@
+	/* Find number of storage in use. */
+	for (int i = 0; i < SMS_BUFFER_ARRAY_SIZE; i++) {
+		if (sms_buffers[i].pdu_size != 0) {
+			sms_buffers_used++;
+		}
+	}
+	return sms_buffers_used;
+}
+
+/* Get internal callback. */
+static nrf_modem_at_cmd_handler_t filter_cb_get(char *at_cmd)
+{
+	bool match;
+
+	for (size_t i = 0; i < NUM_FILTERS_INTERNAL; i++) {
+		match = strstr(at_cmd, gcf_sms_filters[i].cmd);
+		if (match && !gcf_sms_filters[i].paused) {
+			return gcf_sms_filters[i].callback;
+		}
+	}
+
+	return NULL;
+}
+
+/* Callback for filter matches in the Modem library.
+ * Implements the internal filter with multiple command support.
+ */
+int gcf_sms_filter_callback(char *buf, size_t len, char *at_cmd)
+{
+	int err = 0;
+	nrf_modem_at_cmd_handler_t callback;
+	char *msg, *msg_rest;
+	char *buf_remaining = buf;
+	size_t len_remaining = len;
+	int len_tmp = 0;
+	int cmss_cmgs_commands_count = 0;
+	int num_cmss_cmgs_commands = 0;
+
+	if (!at_cmd) {
+		return -NRF_EFAULT;
+	}
+
+	/* Count number of CMSS' and GMGS' in the concatenated AT command. */
+	const char *tmp = at_cmd;
+
+	while ((tmp = strstr(tmp, "+CMSS="))) {
+		num_cmss_cmgs_commands++;
+		tmp++;
+	}
+
+	tmp = at_cmd;
+	while ((tmp = strstr(tmp, "+CMGS="))) {
+		num_cmss_cmgs_commands++;
+		tmp++;
+	}
+
+	/* Split and loop on multiple commands. */
+	msg = strtok_r(at_cmd, ";", &msg_rest);
+
+	if (!msg) {
+		err = -NRF_EFAULT;
+		return err;
+	}
+
+	while (msg) {
+		/* Count CMSS and CMGS commands and send CMMS=0 before the last. */
+		if (prefix_match("AT+CMSS=", msg) || prefix_match("AT+CMGS=", msg)) {
+			cmss_cmgs_commands_count++;
+
+			if (cmss_cmgs_commands_count == num_cmss_cmgs_commands) {
+				LOG_DBG("Sending AT+CMMS=0");
+				nrf_modem_at_printf("AT+CMMS=0");
+			}
+		}
+
+		callback = filter_cb_get(msg);
+
+		if (callback != NULL) {
+			/* AT command is filtered. */
+			err = (*callback)(buf_remaining, len_remaining, msg);
+		} else {
+			err = nrf_modem_at_cmd(buf_remaining, len_remaining, msg);
+		}
+
+		if (err != 0) {
+			return err;
+		}
+
+		/* Get the next command. */
+		msg = strtok_r(msg_rest, ";", &msg_rest);
+		/* Add AT for concatenated AT-commands. */
+		if ((msg) && (msg - 2 * sizeof(char) >= at_cmd)) {
+			msg = msg - 2 * sizeof(char);
+			msg[0] = 'A';
+			msg[1] = 'T';
+		}
+
+		/* Move head of response buffer after current reply. */
+		len_tmp = strlen(buf_remaining);
+		buf_remaining = &(buf_remaining[len_tmp]);
+		len_remaining = len_remaining - len_tmp;
+	}
+
+	return err;
+}
+
+/*
+ * AT Filter callback functions.
+ */
+
+static int at_cmd_callback_cpms(char *buf, size_t len, char *at_cmd)
+{
+	int sms_buffers_used = sms_buffers_num_used_get();
+
+	/* Test */
+	if (prefix_match("AT+CPMS=?", at_cmd)) {
+		return at_custom_cmd_respond(buf, len, "+CPMS: (\"TA\"),(\"TA\")\r\nOK\r\n");
+	}
+
+	/* Set */
+	if (prefix_match("AT+CPMS=\"TA\",\"TA\"", at_cmd)) {
+		return at_custom_cmd_respond(buf, len,
+					      "+CPMS: %d,%d,%d,%d\r\nOK\r\n",
```

Comment:
Nit: there is a small misalignment. In other lines as well.

Reply:
Thanks, updated.